src/main/java/nextstep/courses/service/SessionService.java

```diff
@@ -0,0 +1,86 @@
+package nextstep.courses.service;
+
+import nextstep.courses.domain.image.Image;
+import nextstep.courses.domain.image.ImageRepository;
+import nextstep.courses.domain.session.*;
+import nextstep.courses.domain.student.Student;
+import nextstep.courses.domain.student.StudentRepository;
+import nextstep.payments.domain.Payment;
+import nextstep.users.domain.NsUser;
+import org.springframework.stereotype.Service;
+import org.springframework.transaction.annotation.Transactional;
+
+import javax.annotation.Resource;
+import java.util.List;
+import java.util.stream.Collectors;
+
+@Service("sessionService")
+public class SessionService {
+    @Resource(name = "sessionRepository")
+    private SessionRepository sessionRepository;
+    @Resource(name = "imageRepository")
+    private ImageRepository imageRepository;
+    @Resource(name = "studentRepository")
+    private StudentRepository studentRepository;
+
+    @Transactional(readOnly = true)
+    public Session findById(long sessionId) {
+        Session session = sessionRepository.findById(sessionId);
+        Image image = imageRepository.findBySessionId(sessionId);
+        List<Student> students = studentRepository.findAllBySessionId(sessionId);
+
+        return getSession(session, image, students);
+    }
+
+    private static Session getSession(Session session, Image image, List<Student> students) {
+        if (session.getSessionType().equals(SessionType.FREE)) {
+            return FreeSession.of((FreeSession) session, image, students);
+        }
+        return PaidSession.of((PaidSession) session, image, students);
+    }
+
+    @Transactional(readOnly = true)
+    public List<Session> findAllByCourseId(long courseId) {
+        return sessionRepository.findAllByCourseId(courseId).stream()
+                .map(it -> {
+                    Image image = imageRepository.findBySessionId(it.getId());
+                    List<Student> students = studentRepository.findAllBySessionId(it.getId());
+
+                    return getSession(it, image, students);
+                })
+                .collect(Collectors.toList());
+    }
+
+    @Transactional
+    public long create(Long courseId, Session session, Image image) {
+        long saveSessionId = sessionRepository.save(session, courseId);
+        imageRepository.save(image, saveSessionId);
+
+        return saveSessionId;
+    }
+
+    @Transactional
+    public Session register(Long sessionId, NsUser nsUser, Payment payment) {
+        Session session = sessionRepository.findById(sessionId);
+
+        Registration registration = Registration.of(sessionId, nsUser, payment);
+
+        if (session.getSessionType().equals(SessionType.FREE)) {
+            registerFreeSession((FreeSession) session, registration);
+            return session;
+        }
+        registerPaidSession((PaidSession) session, registration);
```

Comment:
```suggestion
        session.register(registration);
        studentRepository.saveAll(session.getStudents(), session.getId());
```
상속 관계로 설계했으니 그만한 장점도 나와야할 것 같아요 
그러려면 `findById` 시 repository 수준에서 조회된 데이터에 따라 FreeSession 인지 PaidSession인지 반환되어야 할 것 같아요

그러면 서비스 로직에서 타입별 분기 없이 처리할 수 있어보입니다.
어떻게 생각하시나요?

Reply:
이 방법도 고민해 봤었는데요. 깔끔한 구조가 안떠올라서 일단 naive하게 짰었습니다. 혹시, 좋은 방법이 있을까요?

Reply:
> 이 방법도 고민해 봤었는데요. 깔끔한 구조가 안떠올라서 일단 naive하게 짰었습니다. 혹시, 좋은 방법이 있을까요?

Session 테이블에서 모든 컬럼을 조회하고 특정 컬럼값에 따라 특정 Session 구현 객체로 반환하면 될 것 같아요
현재 구조라면 `sessionType` enum 값을 보고 FreeSession인지, PaidSession인지 반환하면 될 것 같네요

JPA에서는 이와 비슷하게 상속 관계 매핑 방법중 단일 테이블 전략을 사용해서 처리할 수 있게 지원하고 있습니다.